Issue#857: MyRocks: Incorrect behavior when multiple statements fail …

spetrunia opened the pull request:
…inside a transaction

rollback_to_stmt_savepoint() calls do_rollback_to_savepoint(). This
removes the changes made by this statement, and also removes the last
set savepoint.

Before we start processing the next statement, we need to set the new
savepoint, so we will have something to rollback to if the next statement
fails.

Since rollback_to_stmt_savepoint always sets a new savepoint now,
m_n_savepoints is now redundant and is removed.

hermanlee commented on mysql-test/suite/rocksdb/r/transaction.result:
```diff
@@ -958,3 +958,20 @@ a
 drop table t1,t2,t3;
+#
+# MDEV-16710: Slave SQL: Could not execute Update_rows_v1 event with RocksDB and triggers
+# Issue#857: MyRocks: Incorrect behavior when miltiple statements fail inside a transaction
```
nit: typo: multiple

hermanlee commented on storage/rocksdb/ha_rocksdb.cc:
```diff
@@ -2436,7 +2436,6 @@ class Rdb_transaction {
       entire transaction.
     */
     do_set_savepoint();
-    m_n_savepoints= 1;
```
Since this is not longer used, the definition in Rdb_transaction should also be removed.

spetrunia replied:
Done

hermanlee reviewed (changes requested):
Can you also update the description with:

Squash with D7509380

On a new rebase, we'll merge this fix with the original change.

spetrunia committed:
Issue#857: MyRocks: Incorrect behavior when miltiple statements fail inside a transaction

rollback_to_stmt_savepoint() calls do_rollback_to_savepoint(). This
removes the changes made by this statement, and also removes the last
set savepoint.

Before we start processing the next statement, we need to set the new
savepoint, so we will have something to rollback to if the next statement
fails.

Since rollback_to_stmt_savepoint always sets a new savepoint now,
m_n_savepoints is now redundant and is removed.

Squash with D7509380

spetrunia commented:
Addressed all review input.
